store/src/java/com/zimbra/qa/unittest/SharedImapTests.java

```diff
@@ -1177,6 +1178,303 @@ public void testCreate() throws Exception {
 
     }
 
+    @Test
+    public void testCopy() throws IOException {
+        connection = connectAndSelectInbox();
+        Flags flags = Flags.fromSpec("afs");
+        for (int i = 0; i < 3; i++) {
+            Date date = new Date(System.currentTimeMillis());
+            Literal msg = message(1000 + i * 1000);
+            try {
+                connection.append("INBOX", flags, date, msg);
+            } finally {
+                msg.dispose();
+            }
+        }
+        connection.create("FOO");
+        connection.copy("1:3", "FOO");
+        connection.select("FOO");
+        Map<Long, MessageData> mdMap = connection.fetch("1:3", "(ENVELOPE INTERNALDATE BODY BODY.PEEK[])");
+        assertEquals("Size of map returned by fetch", 3, mdMap.size());
+    }
+
+    @Test
+    public void testAppendThrottle() throws Exception {
+        connection = connectAndSelectInbox();
+        assertTrue(connection.hasCapability("UIDPLUS"));
+        Date date = new Date(System.currentTimeMillis());
+        Flags flags = Flags.fromSpec("afs");
+        for (int i = 0; i < LOOP_LIMIT; i++) {
+            Literal msg = message(100000);
+            try {
+                connection.append("INBOX", flags, date, msg);
+            } finally {
+                msg.dispose();
+            }
+        }
+
+        Literal msg = message(100000);
+        try {
+            connection.append("INBOX", flags, date, msg);
+            Assert.fail("expected exception here...");
+        } catch (Exception e) {
+            Assert.assertTrue(connection.isClosed());
```

Comment:
General comment.  Best to have a description in asserts so that it is easier when things fail to get a feel for why it has failed without having to check the source.